### PR TITLE
add ws client metrics

### PR DESCRIPTION
**Pull request by akosyakov**
## Description

This PR:
- adds a metrics to monitor state of ws from client perspective
- it adds as well client error reporting if ws errored or closed unexpectedly
- since ws can fail because of issues with a network this PR adds a retry to report metrics and errors
- it also improves generally error reporting for dashboard by putting commit in the bundle and quieting unknown errors for Public API streaming test

<details>
<summary>Summary generated by Copilot</summary>

### <samp>🤖 Generated by Copilot at c27efb6</samp>

This pull request adds metrics reporting for WebSocket connections to the public API endpoint and improves the error reporting mechanism in the dashboard and IDE components. It also introduces a new way to display the client version in the UI using the commit hash from the `config.json` file, which is generated during the build process with `yq`. It modifies several files in the `components` directory, such as `dashboard/BUILD.yaml`, `gitpod-protocol/src/gitpod-service.ts`, `supervisor/frontend/src/ide/ide-web-socket.ts`, and `public-api/typescript/src/metrics.ts`. It also updates the dependencies and configuration for the `supervisor-frontend` and `ide-metrics` components.

</details>

## Related Issue(s)

Fixes #

## How to test

- Open a dashboard, toggle your wifi and wait a bit to break web socket connections.
- Start a workspace, toggle your wifi and wait a bit to break web socket connections.
- Inspect new metrics
```
kubectl port-forward <ide-metrics-pod> 9500:9500
curl localhost:9500/metrics | grep websocket_client
```

You should see something like
```
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 50399    0 50399    0     0  75408      0 --:--:-- --:--:-- --:--:-- 75447
# HELP websocket_client_total Total number of WebSocket connections by the client
# TYPE websocket_client_total counter
websocket_client_total{code="1000",origin="gitpod",status="close",was_clean="1"} 1
websocket_client_total{code="1000",origin="workspace",status="close",was_clean="1"} 5
websocket_client_total{code="1006",origin="gitpod",status="close",was_clean="0"} 2
websocket_client_total{code="unknown",origin="gitpod",status="new",was_clean="unknown"} 10
websocket_client_total{code="unknown",origin="gitpod",status="open",was_clean="unknown"} 9
websocket_client_total{code="unknown",origin="workspace",status="error",was_clean="unknown"} 5
websocket_client_total{code="unknown",origin="workspace",status="new",was_clean="unknown"} 7
websocket_client_total{code="unknown",origin="workspace",status="open",was_clean="unknown"} 2
```

#### Preview status

<p>Gitpod was successfully deployed to your preview environment.</p>
<ul>
	<li><b>🏷️ Name</b> - ak-ws-client-total</li>
	<li><b>🔗 URL</b> - <a href="https://ak-ws-client-total.preview.gitpod-dev.com/workspaces" target="_blank">ak-ws-client-total.preview.gitpod-dev.com/workspaces</a>.</li>
	<li><b>📚 Documentation</b> - See our <a href="https://www.notion.so/gitpod/6debd359591b43688b52f76329d04010#7c1ce80ab31a41e29eff2735e38eec39" target="_blank">internal documentation</a> for information on how to interact with your preview environment.</li>
	<li><b>📦 Version</b> - ak-ws_client_total-gha.17468</li>
	<li><b>🗒️ Logs</b> - <a href="https://console.cloud.google.com/logs/query;query=jsonPayload.kubernetes.host%3D%22preview-ak-ws-client-total%22%0A%0A--%20Filter%20on%20service:%0A--%20jsonPayload.serviceContext.service%3D%22ws-manager-mk2%22%0A;duration=P1D?project=gitpod-core-dev" target="_blank">GCP Logs Explorer</a></li>
</ul>

## Build Options

<details>
<summary>Build</summary>

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`
</details>

<details>
<summary>Publish</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer</summary>

- [ ] analytics=segment
- [ ] with-dedicated-emulation
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

<details>
<summary>Preview Environment / Integration Tests</summary>

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [ ] /werft with-large-vm
- [x] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [ ] with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`. If enabled, `with-preview` and `with-large-vm` will be enabled.
- [x] with-monitoring
</details>

/hold

### CODE REVIEW

File: components/dashboard/src/service/service.tsx
Comment (akosyakov) [inline components/dashboard/src/service/service.tsx:139]: to address noisy unknown error in production

File: components/gitpod-protocol/src/gitpod-service.ts
Comment (akosyakov) [inline components/gitpod-protocol/src/gitpod-service.ts:702]: dead code

File: components/public-api/typescript/src/metrics.ts
Comment (akosyakov) [inline components/public-api/typescript/src/metrics.ts:296]: retrying in case if we were offline last time or any other network issue

File: components/public-api/typescript/src/metrics.ts
Comment (akosyakov) [inline components/public-api/typescript/src/metrics.ts:256]: we could consider preserving them in local storage, but it makes things complex, so won't do it for now

File: components/public-api/typescript/src/metrics.ts
Comment (akosyakov) [inline components/public-api/typescript/src/metrics.ts:438]: here is a concern that there will be some kind of another error not necessary network, but we have not see it so far

File: components/dashboard/BUILD.yaml
Comment (geropl) [inline components/dashboard/BUILD.yaml:28]: nit: Why not make it part of `yarn build` as well? And wrap it into `yarn generate` (or similar)?

File: components/dashboard/src/service/metrics.ts
Comment (geropl) [inline components/dashboard/src/service/metrics.ts:11]: Nice!

File: components/dashboard/src/service/service.tsx
Comment (geropl) [inline components/dashboard/src/service/service.tsx:33]: nit: Can we wrap this into a function `instrumentWebSocketConnection` below?
I feel this file is getting a bit messy lately :sweat_smile:

File: components/public-api/typescript/src/metrics.ts
Comment (geropl) [inline components/public-api/typescript/src/metrics.ts:299]: WDYT about putting an `await` here, as a measure to not overload the backend with too many errors at once?

Comment (akosyakov) [reply components/public-api/typescript/src/metrics.ts:299]: in theory we should also aggregate counters and histograms instead of pushing so many requests

**Comment by geropl** [comment]
@akosyakov LGTM so far. Do you want me to test before approval?

**Comment by akosyakov** [comment]
@geropl I need time to test myself first 😄 for sure would like to have your test afterwards 🙏

File: components/dashboard/src/service/service.tsx
Comment (akosyakov) [inline components/dashboard/src/service/service.tsx:63]: we have to instrument the underlying web socket, not reconnecting, because reconneting never fires error or close and it will disturb data

**Comment by akosyakov** [comment]
@geropl I pushed my changes, after the build it should be alright to test.

**Comment by akosyakov** [comment]
There is one thing which concerns me: when a workspace stopped right now web sockets are always reported like closed abruptly. It would be nice somehow either close them nicely, or indicate a workspace phase. We could add a phase to the metric and error reporting. @geropl wdyt?

**Comment by geropl** [comment]
Just started to look at this again :eyes: 

> It would be nice somehow either close them nicely, or indicate a workspace phase

Sounds good! Workspace phase sounds straight forward; any potential downsides to the client close? I assume you mean on tab close that websockets are interrupted, right? Or on all workspaces stops, because we navigate away? :thinking:

**Comment by akosyakov** [comment]
> Sounds good! Workspace phase sounds straight forward; any potential downsides to the client close? I assume you mean on tab close that websockets are interrupted, right? Or on all workspaces stops, because we navigate away? 🤔

I mean if a workspace get stopped, i.e. with `gp stop`.

**Comment by akosyakov** [comment]
I wonder that the root cause maybe because VS Code Server on SIGTERM does not close ongoing web socket connections, so they get terminated by a pod kill eventually. cc @jeanp413 sounds right?

File: components/public-api/typescript/src/metrics.ts
Comment (akosyakov) [reply components/public-api/typescript/src/metrics.ts:258]: We can add them now probably, I'm just concernted that localstorage is small and it can grow, maybe we put it in indexed db, not sure.

Annother problem is a race we should make it sure that between all windows it is reported only once.

**Review by geropl — APPROVED** [review]
Code LGTM, tested and works :heavy_check_mark:

**Comment by akosyakov** [comment]
got it
```
websocket_client_total{code="1006",instance_phase="stopping",origin="workspace",status="close",was_clean="0"} 2
```

**Pull request merged**
